corgi/core/models.py

```diff
@@ -513,7 +521,78 @@ def get_ofuri(self) -> str:
+        doc.creation_info.set_created_now()
+        for component in self.get_latest_components():
+            comp_spdx_id = self.add_manifest_package(component, doc)
+            # TODO fix the SPDX relationship types to use RelationType class once
+            #  https://github.com/spdx/tools-python/pull/207 is fixed
+            self.add_relationships(
+                comp_spdx_id,
+                component,
+                doc,
+                ComponentNode.ComponentNodeType.PROVIDES,
+                "CONTAINED_BY",
+            )
+            self.add_relationships(
+                comp_spdx_id,
+                component,
+                doc,
+                ComponentNode.ComponentNodeType.PROVIDES_DEV,
+                "DEV_DEPENDENCY_OF",
+            )
+
+        out = StringIO()
+        messages = ErrorMessages()
+        doc.validate(messages)
+        if messages:
+            logger.debug("Manifest document for stream %s is Invalid", self.name)
+            for message in messages:
+                logger.debug(message)
+        # TODO fix validation errors related to pg_verif_code being None (see test_manifests.py)
+        write_document(doc, out, validate=False)
+        return out.getvalue()
+
+    def add_relationships(self, comp_spdx_id, component, doc, component_type, relationship_type):
+        for descendant in component.cnodes.first().get_descendants().filter(type=component_type):
+            prov_spdx_id = self.add_manifest_package(descendant.obj, doc)
+            relationship = Relationship(f"{prov_spdx_id} {relationship_type} {comp_spdx_id}")
+            doc.add_relationships(relationship)
+
+    def add_manifest_package(self, component, doc):
+        # SPDX identifiers can only contain letters number '.' or '-'
+        # We don't just use uuid because we want it to be readable by a human
+        id_list = [component.type, component.name, component.version]
+        if component.arch:
+            id_list.append(component.arch)
+        id_list.append(str(component.uuid)[:8])
+        package = Package(
+            spdx_id="SPDXRef-" + "-".join(id_list),
+            name=component.name,
+            version=f"{component.version}-{component.release}",
+        )
+        external_package_ref = ExternalPackageRef(
+            category="PACKAGE-MANAGER", locator=component.purl, pkg_ext_ref_type="purl"
+        )
+        package.add_pkg_ext_refs(external_package_ref)
+        # TODO add this once https://github.com/spdx/tools-python/issues/193 is fixed
+        #  package.files_analyzed = False
+        package.download_location = "https://access.redhat.com/downloads"
+        if component.copyright_text:
+            package.cr_text = component.copyright_text
+        else:
+            package.cr_text = NoAssert()
+        # TODO need translate license into SPDX format
+        # https://spdx.org/licenses/
+        package.conc_lics = NoAssert()
+        package.license_declared = NoAssert()
```

Comment:
These should already be in SPDX format. The model field license_declared_raw is the raw string from Brew, then the model property license_declared is the same data translated into an SPDX-compliant license expression.